### DISCUSSION TITLE
Hide key value field value on edit page?

### DISCUSSION THREAD

**Discussion by madsem** [filamentphp/filament#2150] [Help] [answered]
My application allows users to enter sensitive data for API connections, atm I'm using a `json` column to store this data in encrypted format.

Is it possible to only hide the value portion of this field on edit screens?

The alternative would be to have multiple nullable columns in the model I guess...

**Comment by danharrin** [comment, marked as answer]
You could hide the entire thing on edit screens using:

```php
->hidden(fn ($livewire) => $livewire instanceof EditRecord)
```

But since it sounds like you want to hide the encrypted data and still allow users to update it, maybe try:

```php
KeyValue::make('FIELD_NAME')
    ->afterStateHydrated(function (KeyValue $component, $livewire, $state): void {
        if ($livewire instanceof CreateRecord) {
            return;
        }

        $component->state(array_map(fn () => '', $state ?? [])); // clear all values
    })
    ->dehydrateStateUsing(function ($state, ?Model $record): array {
        $oldValues = $record?->FIELD_NAME ?? [];

        return array_map(
            fn ($value, $key) => filled($value) ? encrypt($value) : ($oldValues[$key] ?? null),
            $state,
        );
    })
```

This handles automatic encryption of new values for you, and preservation of old ones that have already been saved

might need a little tweaking as I haven't tested it

**Reply by madsem** [reply]
wow, thanks for this! Appreciate you! 
You've hit the nail on the head also, that's exactly what I am trying to accomplish so that users are able to update API keys etc. whenever needed, without exposing the old credentials.

Thanks again for making filament, it's amazing and for most things I'm finding my way around much faster than with Nova.

**Reply by madsem** [reply]
@danharrin been testing it, the `array_map` inside `dehydrateStateUsing` callback is throwing invalid argument count error.
Then I got to think, wouldn't it be much cleaner if this could be done in the `EditAccount` resource.

Otherwise you'd need the above code for every field in that resource that needs this functionality...

The only issue is that the `mutateFormData...` methods have no access to the current Model...
That might be a good idea for a PR?


I.E: If this would be possible, it would allow for a lot cleaner resources when data needs to be mutated a lot:
```php
<?php

namespace App\Filament\Resources\Accounts\AccountResource\Pages;

use Filament\Resources\Pages\EditRecord;
use App\Filament\Resources\Accounts\AccountResource;

class EditAccount extends EditRecord
{

    protected static string $resource = AccountResource::class;


    protected function mutateFormDataBeforeFill(?Model $model, array $data): array
    {
        $data['user_id'] = auth()->id();

        // clear values in api_authentication field to not leak sensitive data
        $data['api_authentication'] = array_map(fn() => '', $data['api_authentication'] ?? []);

        return $data;
    }


    protected function mutateFormDataBeforeSave(?Model $model, array $data): array
    {
        // get old values for this model

        logger($data);
        return $data;
    }
}
```

**Reply by shirokovnv** [reply]
> ```
> ->hidden(fn ($livewire) => $livewire instanceof EditRecord)
> ```

Thanks a lot for the answer!

**Comment by pionl** [comment]
Hi @danharrin, is it possible to not use `hidden` and set "allowed" list of columns in liveware components?

I want to prevent mistakes in future that someone would forget to add something to `$hidden` and leak the data.

In `resource` it was easy: 

```
    public static function getEloquentQuery(): Builder
    {
        // Do not leak data
        return parent::getEloquentQuery()->select([
            User::AttributeId,
            User::AttributeUuid,
            User::AttributeName,
            User::AttributeUsername,
            User::AttributeEmail,
            User::AttributeEmailVerifiedAt,
            User::CREATED_AT,
            User::UPDATED_AT,
        ]);
    }
```

But cant find a way to prevent modal edit and other actions to limit the `record` being loaded with all fields.

thank you

**Reply by pionl** [reply]
I've managed to get to underline liveware logic ModelSynth and `newQueryForRestoration` on the model.

I don't see an option to prevent loading all the date for specific component :(
